quickwit/quickwit-search/src/leaf.rs

```diff
@@ -446,88 +542,134 @@ pub async fn leaf_search(
+            })
+        }
+    }
 
-    // Merging is a cpu-bound task.
-    // It should be executed by Tokio's blocking threads.
-    let span = info_span!("merge_search_responses");
-    let mut merged_search_response = crate::run_cpu_intensive(move || {
-        let _span_guard = span.enter();
-        merge_collector.merge_fruits(split_search_responses)
-    })
-    .await
-    .context("failed to merge split search responses")??;
+    crate::run_cpu_intensive(|| incremental_merge_collector.finalize().map_err(Into::into))
+        .instrument(info_span!("incremental_merge_finalize"))
+        .await
+        .context("failed to merge split search responses")?
+}
 
-    merged_search_response
-        .failed_splits
-        .extend(errors.into_iter().map(|(split_id, err)| SplitSearchError {
-            split_id,
-            error: format!("{err}"),
-            retryable_error: true,
-        }));
-    Ok(merged_search_response)
+#[allow(clippy::too_many_arguments)]
+fn launch_leaf_single_split(
+    request: SearchRequest,
+    searcher_context: Arc<SearcherContext>,
+    index_storage: Arc<dyn Storage>,
+    doc_mapper: Arc<dyn DocMapper>,
+    split: SplitIdAndFooterOffsets,
+    split_filter: Arc<Mutex<CanSplitDoBetter>>,
+    incremental_merge_collector: Arc<Mutex<IncrementalCollector>>,
+    leaf_split_search_permit: tokio::sync::OwnedSemaphorePermit,
+) -> tokio::task::JoinHandle<()> {
+    tokio::spawn(
+        async move {
+            crate::SEARCH_METRICS.leaf_searches_splits_total.inc();
+            let timer = crate::SEARCH_METRICS
+                .leaf_search_split_duration_secs
+                .start_timer();
+            let leaf_search_single_split_res = leaf_search_single_split(
+                &searcher_context,
+                request,
+                index_storage,
+                split.clone(),
+                doc_mapper,
+            )
+            .await;
+            if leaf_search_single_split_res.is_ok() {
+                timer.observe_duration();
+            }
+            let mut locked_incremental_merge_collector =
+                incremental_merge_collector.lock().unwrap();
+            match leaf_search_single_split_res {
+                Ok(split_search_res) => {
+                    locked_incremental_merge_collector.add_split(split_search_res)
+                }
+                Err(err) => locked_incremental_merge_collector.add_failed_split(SplitSearchError {
+                    split_id: split.split_id.clone(),
+                    error: format!("{err}"),
+                    retryable_error: true,
+                }),
+            }
+            if let Some(last_hit) = locked_incremental_merge_collector.peek_worst_hit() {
+                split_filter.lock().unwrap().record_new_worst_hit(last_hit);
+            }
+            // We explicitly drop it, to highlight it to the reader and to force the move.
+            std::mem::drop(leaf_split_search_permit);
```

Comment:
Maybe this is playing with fire, but couldn't we drop that lock before appending to the merge collector?